File: python/aide_frame/web_request.py
"""
HTTP request utilities for server-side web requests.

This module provides simple functions for making HTTP requests from the server,
useful for fetching data from external APIs.

Example usage:
    from aide_frame.web_request import fetch_json, fetch_text

    # Fetch JSON from API
    data = fetch_json("https://api.example.com/data")
    if data:
        print(data["result"])

    # Fetch with custom headers
    data = fetch_json(
        "https://api.example.com/protected",
        headers={"Authorization": "Bearer token123"}
    )
"""

import urllib.request
import urllib.error
import json
from .log import logger


def fetch_json(url, headers=None, timeout=10):
    """
    Fetch and parse JSON from a URL.

    Args:
        url: The URL to fetch from
        headers: Optional dict of HTTP headers
        timeout: Request timeout in seconds (default: 10)

    Returns:
        Parsed JSON as dict/list, or None on error
    """
    try:
        req = urllib.request.Request(url)
        if headers:
            for key, value in headers.items():
                req.add_header(key, value)
        req.add_header('User-Agent', 'aide-frame/1.0')

        logger.debug(f"HTTP: GET {url}")
        with urllib.request.urlopen(req, timeout=timeout) as response:
            body = response.read().decode('utf-8')
            data = json.loads(body)
            # Log response preview (first 100 chars)
            preview = body[:100] + '...' if len(body) > 100 else body
            logger.debug(f"  -> {response.status} {preview}")
            return data
    except urllib.error.HTTPError as e:
        logger.error(f"HTTP error fetching {url}: {e.code} {e.reason}")
        return None
    except urllib.error.URLError as e:
        logger.error(f"URL error fetching {url}: {e.reason}")
        return None
    except json.JSONDecodeError as e:
        logger.error(f"JSON decode error from {url}: {e}")
        return None
    except Exception as e:
        logger.error(f"Unexpected error fetching {url}: {e}")
        return None


def fetch_text(url, headers=None, timeout=10):
    """
    Fetch text content from a URL.

    Args:
        url: The URL to fetch from
        headers: Optional dict of HTTP headers
        timeout: Request timeout in seconds (default: 10)

    Returns:
        Response body as string, or None on error
    """
    try:
        req = urllib.request.Request(url)
        if headers:
            for key, value in headers.items():
                req.add_header(key, value)
        req.add_header('User-Agent', 'aide-frame/1.0')

        logger.debug(f"HTTP: GET {url}")
        with urllib.request.urlopen(req, timeout=timeout) as response:
            body = response.read().decode('utf-8')
            # Log response preview (first 100 chars)
            preview = body[:100] + '...' if len(body) > 100 else body
            logger.debug(f"  -> {response.status} {preview}")
            return body
    except urllib.error.HTTPError as e:
        logger.error(f"HTTP error fetching {url}: {e.code} {e.reason}")
        return None
    except urllib.error.URLError as e:
        logger.error(f"URL error fetching {url}: {e.reason}")
        return None
    except Exception as e:
        logger.error(f"Unexpected error fetching {url}: {e}")
        return None

File: js/aide_frame/src/web-request.js
/**
 * HTTP request utilities for server-side web requests.
 *
 * This module provides simple functions for making HTTP requests from the server,
 * useful for fetching data from external APIs.
 *
 * Example usage:
 *   const { fetchJson, fetchText } = require('aide-frame').webRequest;
 *
 *   // Fetch JSON from API
 *   const data = await fetchJson("https://api.example.com/data");
 *   if (data) {
 *       console.log(data.result);
 *   }
 *
 *   // Fetch with custom headers
 *   const data = await fetchJson(
 *       "https://api.example.com/protected",
 *       { headers: { "Authorization": "Bearer token123" } }
 *   );
 */

const { logger } = require('./log');

/**
 * Fetch and parse JSON from a URL.
 *
 * @param {string} url - The URL to fetch from
 * @param {object} options - Options object
 * @param {object} options.headers - Optional HTTP headers
 * @param {number} options.timeout - Request timeout in milliseconds (default: 10000)
 * @returns {Promise<object|null>} Parsed JSON as object/array, or null on error
 */
async function fetchJson(url, options = {}) {
    const { headers = {}, timeout = 10000 } = options;

    try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), timeout);

        const fetchHeaders = {
            'User-Agent': 'aide-frame/1.0',
            ...headers,
        };

        logger.debug(`HTTP: GET ${url}`);
        const response = await fetch(url, {
            headers: fetchHeaders,
            signal: controller.signal,
        });
        clearTimeout(timeoutId);

        if (!response.ok) {
            logger.error(`HTTP error fetching ${url}: ${response.status} ${response.statusText}`);
            return null;
        }

        const text = await response.text();
        const data = JSON.parse(text);

        // Log response preview (first 100 chars)
        const preview = text.length > 100 ? text.substring(0, 100) + '...' : text;
        logger.debug(`  -> ${response.status} ${preview}`);

        return data;
    } catch (e) {
        if (e.name === 'AbortError') {
            logger.error(`Timeout fetching ${url}`);
        } else if (e instanceof SyntaxError) {
            logger.error(`JSON decode error from ${url}: ${e.message}`);
        } else {
            logger.error(`Error fetching ${url}: ${e.message}`);
        }
        return null;
    }
}

/**
 * Fetch text content from a URL.
 *
 * @param {string} url - The URL to fetch from
 * @param {object} options - Options object
 * @param {object} options.headers - Optional HTTP headers
 * @param {number} options.timeout - Request timeout in milliseconds (default: 10000)
 * @returns {Promise<string|null>} Response body as string, or null on error
 */
async function fetchText(url, options = {}) {
    const { headers = {}, timeout = 10000 } = options;

    try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), timeout);

        const fetchHeaders = {
            'User-Agent': 'aide-frame/1.0',
            ...headers,
        };

        logger.debug(`HTTP: GET ${url}`);
        const response = await fetch(url, {
            headers: fetchHeaders,
            signal: controller.signal,
        });
        clearTimeout(timeoutId);

        if (!response.ok) {
            logger.error(`HTTP error fetching ${url}: ${response.status} ${response.statusText}`);
            return null;
        }

        const text = await response.text();

        // Log response preview (first 100 chars)
        const preview = text.length > 100 ? text.substring(0, 100) + '...' : text;
        logger.debug(`  -> ${response.status} ${preview}`);

        return text;
    } catch (e) {
        if (e.name === 'AbortError') {
            logger.error(`Timeout fetching ${url}`);
        } else {
            logger.error(`Error fetching ${url}: ${e.message}`);
        }
        return null;
    }
}

module.exports = {
    fetchJson,
    fetchText,
};
